pkg/controller/remotemachineset/openstackactuator.go

```diff
@@ -134,3 +156,46 @@ func decodeOpenStackMachineProviderSpec(rawExt *runtime.RawExtension, scheme *ru
 	}
 	return spec, nil
 }
+
+// yamlOptsBuilder lets us provide our own functions to return a 'clouds.yaml' file that has been
+// unmarshaled into the format expected by the OpenStack clients.
+type yamlOptsBuilder struct {
+	cloudYaml map[string]clientconfig.Cloud
+}
+
+func newYamlOptsBuilder(kubeClient client.Client, credsSecretKey types.NamespacedName) (*yamlOptsBuilder, error) {
+
+	credsSecret := &corev1.Secret{}
+	if err := kubeClient.Get(context.TODO(), credsSecretKey, credsSecret); err != nil {
+		return nil, err
+	}
+
+	cloudsYaml, ok := credsSecret.Data[constants.OpenStackCredentialsName]
+	if !ok {
+		return nil, errors.New("did not find credentials in the ClusterDeployment's secret")
+	}
+
+	var clouds clientconfig.Clouds
+	if err := yaml.Unmarshal(cloudsYaml, &clouds); err != nil {
```

Comment:
How stable is `clientconfig.Clouds`? Do you anticipate any issues with supporting older OpenShift versions that may use an older version of this type?

Reply:
This is a good question. Outside of this PR I've never dealt with this package before.
Now that you ask, I would worry not so much about versions of OpenShift but the version of OpenStack that this package will be making calls into. Is there someone more familiar with OpenStack that can perhaps answer this?

As far as OpenShift releases, older versions of the installer made you specify whether this "trunk" support was available, and now it is dynamic. And we mark the field as optional in Hive https://github.com/openshift/hive/blob/master/pkg/apis/hive/v1/openstack/platform.go#L19. This feels like we've just automated detection of what you previously had to specify.